<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro">
  <property name="PI" value="3.1415926535897931" />
  
  
  <material name="servo">
    <color rgba="0.5 0.5 0.5 1"/>
  </material>

  <material name="support">
    <color rgba="1 0 0 1"/>
  </material>
  
  <material name="sensor">
    <color rgba="0 1 0 1"/>
  </material>



  

  <xacro:macro name="create_box" params="link_name mass size_x size_y size_z material collision_name='' *origin">
    <link name="${link_name}">
      <visual>
        <geometry>
          <box size="${size_x} ${size_y} ${size_z}"/>
        </geometry>
        <xacro:insert_block name="origin" />
        <material name="${material}"/>
      </visual>
      
      <collision name="${collision_name}">
        <geometry>
          <box size="${size_x} ${size_y} ${size_z}"/>
        </geometry>
        <xacro:insert_block name="origin" />
      </collision>
      
      <inertial>
        <xacro:insert_block name="origin" />
        <mass value="${mass}"/>
        <inertia
          ixx="${mass*(size_y*size_y + size_z*size_z)/12}" ixy="0.0" ixz="0.0"
          iyy="${mass*(size_x*size_x + size_z*size_z)/12}" iyz="0.0"
          izz="${mass*(size_x*size_x + size_y*size_y)/12}"/>
      </inertial>
    </link>
  </xacro:macro>
  
  <xacro:macro name="create_cylinder" params="link_name mass radius length material collision_name='' *origin">
    <link name="${link_name}">
      <visual>
        <geometry>
          <cylinder length="${length}" radius="${radius}"/>
        </geometry>
        <xacro:insert_block name="origin" />
        <material name="${material}"/>
      </visual>
      
      <collision name="${collision_name}">
        <geometry>
          <cylinder length="${length}" radius="${radius}"/>
        </geometry>
        <xacro:insert_block name="origin" />
      </collision>
      
      <inertial>
        <xacro:insert_block name="origin" />
        <mass value="${mass}"/>
        <inertia
          ixx="${mass*radius*radius/2}" ixy="0.0" ixz="0.0"
          iyy="${mass*radius*radius/2}" iyz="0.0"
          izz="0.0"/>
      </inertial>
    </link>
  </xacro:macro>

  <xacro:macro name="create_sphere" params="link_name mass radius material collision_name='' *origin">
    <link name="${link_name}">
      <visual>
        <geometry>
          <sphere radius="${radius}"/>
        </geometry>
        <xacro:insert_block name="origin" />
        <material name="${material}"/>
      </visual>
      
      <collision name="${collision_name}">
        <geometry>
          <sphere radius="${radius}"/>
        </geometry>
        <xacro:insert_block name="origin" />
      </collision>
      
      <inertial>
        <xacro:insert_block name="origin" />
        <mass value="${mass}"/>
        <inertia
          ixx="${2*mass*radius*radius/5}" ixy="0.0" ixz="0.0"
          iyy="${2*mass*radius*radius/5}" iyz="0.0"
          izz="${2*mass*radius*radius/5}"/>
      </inertial>
    </link>
  </xacro:macro>


  


  <xacro:macro name="single_leg" params="parent_link *origin leg_name">
    <xacro:create_box
      link_name="${leg_name}_base_link"
      size_x="0.05"
      size_y="0.057"
      size_z="0.025"
      mass="0.0455"
      material="servo"
      >
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </xacro:create_box>
    
    <joint name="${leg_name}_base_hip" type="revolute">
      <parent link="${parent_link}"/>
      <child link="${leg_name}_base_link"/>
      <axis xyz="1 0 0"/>
      <xacro:insert_block name="origin" />
      <limit lower="-${PI/2}" upper="${PI/2}" velocity="10" effort="100"/>
    </joint>


    <xacro:create_box
      link_name="${leg_name}_hip_servo"
      size_x="0.057"
      size_y="0.055"
      size_z="0.025"
      mass="0.0455"
      material="servo"
      >
        <origin rpy="0 0 0" xyz="0 0 -0.025"/>
    </xacro:create_box>
    
    <joint name="${leg_name}_hip_servos_joint" type="fixed">
      <parent link="${leg_name}_base_link"/>
      <child link="${leg_name}_hip_servo"/>
      <origin xyz="-0.013 0 0"/>
    </joint>


    <xacro:create_box
      link_name="${leg_name}_thigh"
      size_x="0.025"
      size_y="0.055"
      size_z="0.136"
      mass="0.01"
      material="support"
      >
        <origin rpy="0 0 0" xyz="0 0 -0.05525"/>
    </xacro:create_box>

    <joint name="${leg_name}_thigh_lift" type="revolute">
      <parent link="${leg_name}_hip_servo"/>
      <child link="${leg_name}_thigh"/>
      <axis xyz="0 1 0"/>
      <origin xyz="0.0095 0 -0.0265"/>
      <limit lower="-${PI/2}" upper="${PI/2}" velocity="10" effort="100"/>
    </joint>


    <xacro:create_box
      link_name="${leg_name}_knee_servo"
      size_x="0.025"
      size_y="0.055"
      size_z="0.061"
      mass="0.0455"
      material="servo"
      >
        <origin rpy="0 0 0" xyz="0 0 -0.01095"/>
    </xacro:create_box>

    <joint name="${leg_name}_knee_lift" type="revolute">
      <parent link="${leg_name}_thigh"/>
      <child link="${leg_name}_knee_servo"/>
      <axis xyz="0 1 0"/>
      <origin xyz="0 0 -0.114"/>
      <limit lower="0" upper="${PI}" velocity="10" effort="100"/>
    </joint>
    
    
    <xacro:create_cylinder
      link_name="${leg_name}_toe"
      length="0.087"
      radius="0.0065"
      mass="0.01"
      material="support"
      >
        <origin rpy="0 0 0" xyz="0 0 -0.01095"/>
    </xacro:create_cylinder>
    
    <joint name="${leg_name}_knee_to_toe" type="fixed">
      <parent link="${leg_name}_knee_servo"/>
      <child link="${leg_name}_toe"/>
      <origin xyz="0 0 -0.0425"/>
    </joint>
    
    
    <xacro:create_sphere
      link_name="${leg_name}_contact_sensor"
      radius="0.0065"
      mass="0.0"
      material="sensor"
      collision_name="${leg_name}_cs"
      >
        <origin rpy="0 0 0" xyz="0 0 -0.054"/>
    </xacro:create_sphere>
    
    <joint name="${leg_name}_contact_sensor_joint" type="fixed">
      <parent link="${leg_name}_toe"/>
      <child link="${leg_name}_contact_sensor"/>
      <!--<origin xyz="0 0 -0.114"/>-->
    </joint>

    
    <gazebo reference="${leg_name}_contact_sensor">
        <!-- Add grip to the rubber -->
        <kp>100000.0</kp>
        <kd>100000.0</kd>
        <mu1>10.0</mu1>
        <mu2>0.1</mu2>
        <fdir1>1 0 0</fdir1>
        <material>Gazebo/Red</material>

        <sensor type="contact" name="${leg_name}_contact">
            <contact>
                <collision>${leg_name}_knee_servo_fixed_joint_lump__${leg_name}_cs_collision_2</collision>
            </contact>
        </sensor>
    </gazebo>

  </xacro:macro>
</robot>

